sanic/websocket.py

```diff
@@ -7,23 +10,748 @@
     MutableMapping,
     Optional,
     Union,
+    Iterable,
+    Sequence,
+    Mapping, AsyncIterator
 )
+import codecs
 
 from httptools import HttpParserUpgrade  # type: ignore
-from websockets import (  # type: ignore
-    ConnectionClosed,
-    InvalidHandshake,
-    WebSocketCommonProtocol,
-    handshake,
-)
+from websockets.server import ServerConnection
+from websockets.connection import Event, OPEN, CLOSING, CLOSED
+from websockets.exceptions import ConnectionClosed, InvalidHandshake, InvalidOrigin, InvalidUpgrade, \
+    ConnectionClosedError
+from websockets.typing import Data
+from websockets.frames import Frame, Opcode, prepare_ctrl, OP_PONG
+from websockets.utils import accept_key
 
-from sanic.exceptions import InvalidUsage
-from sanic.server import HttpProtocol
+from sanic.exceptions import InvalidUsage, Forbidden, SanicException
+from sanic.server import HttpProtocol, SanicProtocol
+from sanic.log import error_logger, logger
 
+import asyncio
 
-__all__ = ["ConnectionClosed", "WebSocketProtocol", "WebSocketConnection"]
 
 ASIMessage = MutableMapping[str, Any]
+UTF8Decoder = codecs.getincrementaldecoder("utf-8")
+
+class WebsocketFrameAssembler:
```

Comment:
FYI there's a strong chance I end up adding a version of this to websockets, when I write an asyncio implementation on top of the Sans I/O layer. I'll document it so you can reuse it!

Reply:
Which reminds me that Sanic has been using non-incremental (= buggy) UTF-8 decoding in some example code in docs and also in tests. Hopefully this got cleared when someone rewrote the manual a few months ago but checking to make sure wouldn't be a bad idea. Tests don't hit the bug because they only use ASCII strings.